scripts/syntactics/common/utils.py

```diff
@@ -0,0 +1,424 @@
+    W = orthonormal_initializer(lstm_hiddens, lstm_hiddens + input_dims, debug)
+    W_h, W_x = W[:, :lstm_hiddens], W[:, lstm_hiddens:]
+    b = nd.zeros((4 * lstm_hiddens,))
+    b[lstm_hiddens:2 * lstm_hiddens] = -1.0
+    lstm_cell = rnn.LSTMCell(input_size=input_dims, hidden_size=lstm_hiddens,
+                             i2h_weight_initializer=mx.init.Constant(np.concatenate([W_x] * 4, 0)),
+                             h2h_weight_initializer=mx.init.Constant(np.concatenate([W_h] * 4, 0)),
+                             h2h_bias_initializer=mx.init.Constant(b))
+    wrapper = VariationalDropoutCell(lstm_cell, drop_states=dropout_h)
+    return wrapper
+
+
+def biLSTM(f_lstm, b_lstm, inputs, batch_size=None, dropout_x=0., dropout_h=0.):
+    """
+    Feature extraction through BiLSTM
+    :param inputs: # seq_len x batch_size
+    :param batch_size:
+    :return: Outputs of BiLSTM layers, seq_len x 2 hidden_dims x batch_size
+    """
+    for f, b in zip(f_lstm, b_lstm):
+        inputs = nd.Dropout(inputs, dropout_x, axes=[0])  # important for variational dropout
+        fo, fs = f.unroll(length=inputs.shape[0], inputs=inputs, layout='TNC', merge_outputs=True)
+        bo, bs = b.unroll(length=inputs.shape[0], inputs=inputs.flip(axis=0), layout='TNC', merge_outputs=True)
+        f.reset(), b.reset()
+        inputs = nd.concat(fo, bo.flip(axis=0), dim=2)
+    return inputs
+
+
+def leaky_relu(x):
+    """
+    slope=0.1 leaky ReLu
+    :param x: NDArray
+    :return: NDArray
+    """
+    return nd.LeakyReLU(x, slope=.1)
+
+
+def bilinear(x, W, y, input_size, seq_len, batch_size, num_outputs=1, bias_x=False, bias_y=False):
+    """
+    Do xWy
+
+    :param x: (input_size x seq_len) x batch_size
+    :param W: (num_outputs x ny) x nx
+    :param y: (input_size x seq_len) x batch_size
+    :param input_size: input dimension
+    :param seq_len: sequence length
+    :param batch_size: batch size
+    :param num_outputs: number of outputs
+    :param bias_x: whether concat bias vector to input x
+    :param bias_y: whether concat bias vector to input y
+    :return: [seq_len_y x seq_len_x if output_size == 1 else seq_len_y x num_outputs x seq_len_x] x batch_size
+    """
+    if bias_x:
+        x = nd.concat(x, nd.ones((1, seq_len, batch_size)), dim=0)
+    if bias_y:
+        y = nd.concat(y, nd.ones((1, seq_len, batch_size)), dim=0)
+
+    nx, ny = input_size + bias_x, input_size + bias_y
+    # W: (num_outputs x ny) x nx
+    lin = nd.dot(W, x)
+    if num_outputs > 1:
+        lin = reshape_fortran(lin, (ny, num_outputs * seq_len, batch_size))
+    y = y.transpose([2, 1, 0])  # May cause performance issues
+    lin = lin.transpose([2, 1, 0])
+    blin = nd.batch_dot(lin, y, transpose_b=True)
+    blin = blin.transpose([2, 1, 0])
+    if num_outputs > 1:
+        blin = reshape_fortran(blin, (seq_len, num_outputs, seq_len, batch_size))
+    return blin
+
+
+def orthonormal_initializer(output_size, input_size, debug=False):
```

Comment:
I'm not sure if it's the same as the `Orthogonal` initializer in MXNet. https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/initializer.py#L521

Reply:
Not the same, not mentioned in parser paper. I'll replace this with the SVD implementation, such that a black box is removed.